activation/e2e/checkpoint_merged_test.go

```diff
@@ -0,0 +1,348 @@
+
+	mergedIdAtx := createInitialAtx(publish, goldenATX, goldenATX, niposts[1].NIPostState, initialPosts[1])
+	mergedIdAtx.Sign(mergedID)
+
+	marriageATX := createInitialAtx(publish, goldenATX, goldenATX, niposts[0].NIPostState, initialPosts[0])
+	marriageATX.Marriages = []wire.MarriageCertificate{
+		{
+			Signature: mainID.Sign(signing.MARRIAGE, mainID.NodeID().Bytes()),
+		},
+		{
+			ReferenceAtx: mergedIdAtx.ID(),
+			Signature:    mergedID.Sign(signing.MARRIAGE, mainID.NodeID().Bytes()),
+		},
+	}
+	marriageATX.Sign(mainID)
+	logger.Info("publishing marriage ATX", zap.Inline(marriageATX))
+
+	mFetch.EXPECT().RegisterPeerHashes(peer.ID(""), gomock.Any()).Times(2)
+	mFetch.EXPECT().GetPoetProof(gomock.Any(), gomock.Any()).Times(2)
+	mFetch.EXPECT().GetAtxs(gomock.Any(), []types.ATXID{mergedIdAtx.ID()}, gomock.Any())
+	mBeacon.EXPECT().OnAtx(gomock.Any()).Times(2)
+	mTortoise.EXPECT().OnAtx(gomock.Any(), gomock.Any(), gomock.Any()).Times(2)
+	err = atxHdlr.HandleGossipAtx(context.Background(), "", codec.MustEncode(mergedIdAtx))
+	require.NoError(t, err)
+	err = atxHdlr.HandleGossipAtx(context.Background(), "", codec.MustEncode(marriageATX))
+	require.NoError(t, err)
+
+	// Step 2. Publish merged ATX together
+	publish = marriageATX.PublishEpoch + 2
+	eg, ctx = errgroup.WithContext(context.Background())
+	// 2.1. NiPOST for main ID (the publisher)
+	eg.Go(func() error {
+		n, err := buildNipost(ctx, nb, mainID, publish, marriageATX.ID(), marriageATX.ID())
+		logger.Info("built NiPoST", zap.Any("post", n))
+		niposts[0] = n
+		return err
+	})
+
+	// 2.2. NiPOST for merged ID
+	prevATXID, err := atxs.GetLastIDByNodeID(db, mergedID.NodeID())
+	require.NoError(t, err)
+	eg.Go(func() error {
+		n, err := buildNipost(ctx, nb, mergedID, publish, prevATXID, marriageATX.ID())
+		logger.Info("built NiPoST", zap.Any("post", n))
+		niposts[1] = n
+		return err
+	})
+	require.NoError(t, eg.Wait())
+
+	// 2.3 Construct a multi-ID poet membership merkle proof for both IDs
+	_, members, err := poetSvc.Proof(context.Background(), "1")
+	require.NoError(t, err)
+	membershipProof := constructMerkleProof(t, members, map[uint64]bool{0: true, 1: true})
+
+	mergedATX := createMerged(
+		t,
+		niposts[:],
+		publish,
+		marriageATX.ID(),
+		marriageATX.ID(),
+		[]types.ATXID{marriageATX.ID(), prevATXID},
+		membershipProof,
+	)
+	mergedATX.VRFNonce = nonces[0]
+	mergedATX.Sign(mainID)
+
+	// 2.4 Publish
+	<-clock.AwaitLayer(mergedATX.PublishEpoch.FirstLayer())
+	logger.Info("publishing merged ATX", zap.Inline(mergedATX))
+
+	mFetch.EXPECT().RegisterPeerHashes(peer.ID(""), gomock.Any())
+	mFetch.EXPECT().GetPoetProof(gomock.Any(), gomock.Any())
+	mFetch.EXPECT().GetAtxs(gomock.Any(), gomock.Any(), gomock.Any())
+	mBeacon.EXPECT().OnAtx(gomock.Any())
+	mTortoise.EXPECT().OnAtx(gomock.Any(), gomock.Any(), gomock.Any())
+	err = atxHdlr.HandleGossipAtx(context.Background(), "", codec.MustEncode(mergedATX))
+	require.NoError(t, err)
+
+	// Step 3. Checkpoint
```

Comment:
Everything before this I think we don't really need to test e2e - we could just start here with a DB containing all the data that was generated up until this point. The point of this test is to verify that checkpoint is working e2e not the atxbuilder or handler. wdyt?

Reply:
That would work too, but on the other hand, the test would need to assume much about the implementation (it would need to put atxs and marriages into the DB). A fully e2e test using the exposed API surface (ATX handler, ATX wire types) has less chance of breaking in the future when the internal implementation changes.